src/transformers/models/beit/image_processing_beit.py

```diff
@@ -109,12 +108,7 @@ def __init__(
         **kwargs,
     ) -> None:
         if "reduce_labels" in kwargs:
-            warnings.warn(
-                "The `reduce_labels` parameter is deprecated and will be removed in a future version. Please use"
-                " `do_reduce_labels` instead.",
-                FutureWarning,
-            )
-            do_reduce_labels = kwargs.pop("reduce_labels")
+            raise ValueError("The `reduce_labels` parameter has been deprecated. Use `do_reduce_labels` instead.")
```

Comment:
Sadly we can't do this directly for two reasons: 

* A deprecation version wasn't specified in the message (my bad)
* This is going to break a lot of things for users

We absolutely want to push users to try and use the correct argument, and we should update this value on all of the official checkpoints. However, because this is from the config, which users might not have control over or know how to change; and there might be hundreds of configs with the old value on and off the hub, we're at risk of breaking code and losing users.

What I'd suggest here is: 
* We update the warning message to specify a specific version (typically two versions after the release version this commit will be included in). 
* After this version, we remove the warning. We silently handle this to maintain backwards compatibility i.e. popping the value, and make sure all official references to the flag are removed; we don't guarantee any future maintenance for this value.

Reply:
Yes, that's breaking changes...

The idea was to support reading old configs overriding `from_dict` method with
```python
@classmethod
    def from_dict(cls, image_processor_dict: Dict[str, Any], **kwargs):
    ...
    if "reduce_labels" in image_processor_dict:
            image_processor_dict["do_reduce_labels"] = image_processor_dict.pop("reduce_labels")
```
Here we silently replace `reduce_labels` with `do_reduce_labels`.

By raising an exception in `__init__`, we prevent users from creating **new** `image_processor` with deprecated parameters. So that might break some fine-tuning scripts, which the user should have control of.

Does it make sense, or am I missing some use cases?
Are such kinds of breaking changes possible only over major versions like v4->v5?

Reply:
> Here we silently replace reduce_labels with do_reduce_labels.

This is great - we definitely want to do this! It means any old config, if saved out, will be fully compatible. 

> Are such kinds of breaking changes possible only over major versions like v4->v5?

Kind of, it's more to do with the type and scale of breaking change, and the controls users have over it. 

We do deprecate methods, flags, etc. within the codebase semi-regularly. And we do tolerate a certain amount of breaking changes. For example, in musicgen recently, training was enabled. This technically created a breaking change, as one of the models previously returned a loss. The update meant that anyone using the model would see the loss suddenly change. However, as the previous loss was incorrect and the model use relatively low this was deemed OK. 

There's two main reasons why we wouldn't want to raise an exception here at the moment:
* Ability to change the code behaviour. Users don't always have the option of updating the config: they might not have permissions to modify on the hub.
* Awareness of configs. Many users are not familiar with the configs. It's easier to deprecate calls to e.g. `object.method(foo, bar)` as we emit a warning, and the user just removes that line of code. Changing the config is less obvious. 

The reason I think this is because when we first tried to deprecate flags for the image processors, we got quite a few messages from users either complaining or confused about why they were happening. I'm yet to see a single config updated because of them!

It is something we could do in v5, although we may not want to.

Reply:
Thanks a lot for the clarification, I will summarize PR just to make sure we are on the same page.

There are 3 most popular ways to create an image processor with deprecated parameters (sorted by popularity, high->low, IMHO)

| Method                                                                                                                        | Is BC in current PR                    |
|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------|
| image_processor = ImageProcessor.from_pretrained(<br>&emsp;    "repo/model_with_deprecated_params"<br>)                             | ✅ Yes<br>Because the user does not<br> have control of it (by overriding `from_dict()`)          |
| image_processor = ImageProcessor.from_pretrained(<br>&emsp;    "repo/model_with_deprecated_params",<br>&emsp;    reduce_labels=True,<br>) | 🟥 No, raise error<br>Because it is explicit parameter setting |
| image_processor = ImageProcessor(<br>&emsp;    ...<br>&emsp;    reduce_labels=True,<br>)                                                  | 🟥 No, raise error<br>Because it is explicit parameter setting |
|                                                                                                                               |                                               |

I hope this approach:
1. Pushes the user to update their codebase, which they have control over, while still allowing using previously created models.
2. Will allow us to gradually move backward compatibility for deprecated params only in one method `from_dict` (ideally), making code easier to support.
3. Probably, we can also get rid of `**kwargs`, have a strict set of arguments, and replace checks with a decorator for renamed/removed arguments:
```python
@depecated("reduce_labels", replace_with="do_reduce_labels", from="4.42.0", to="4.45.0")
def preprocess(...):
```
The decorator will rename the argument and raise a warning in between, and then we can remove it with automated checks, or leave it for a while to raise an explicit error.
At the moment, because of `**kwargs`, anyone might easily make a typo in an argument name and it will be silently passed, leading to unexpected behavior that is hard to debug.
As a simple example:
```python
from transformers import DetrImageProcessor

image_processor = DetrImageProcessor.from_pretrained(
    "facebook/detr-resnet-50",
    do_nomalize=False, <----- here set as False
)
print(image_processor)

# DetrImageProcessor {
# ...
#   "do_convert_annotations": true,
#   "do_normalize": true, <----- but we get True
#   "do_pad": true, 
#   "do_rescale": true,
#   "do_resize": true,
# ...
# }
```

We can test this strategy with some particular models, not super popular, just to see how it works.

I might overestimate the problem or underestimate risks, user behavior, or edge cases. You have much more experience maintaining the library, so let me know if any of these thoughts are valuable. And thank you once again for your clarifications with examples, it is much easier to follow 🤗

Reply:
100% aligned with you comment above, thanks for taking the time to write this up so clearly. 

I really like the idea of a decorator. Let's ask @molbap his opinion on this, as he's been handling a lot of these considerations across processors and image processor.

Reply:
OOO for a few days but checked and it's an interesting take. I like the decorator idea too. I wholeheartedly agree with
>  At the moment, because of **kwargs, anyone might easily make a typo in an argument name and it will be silently passed, leading to unexpected behavior that is hard to debug.

That is almost always true. That's why getting rid of **kwargs is impossible, as we need to be able to handle arbitrary inputs without breaking runtime in situations where people have non-canonical configs.

Hence the point of having a list of valid arguments: that way, we can at least raise a warning and list the arguments passed by a user that are not doing anything. In your case, if `do_normalize` is part of the valid image processor arguments but `do_nomalize` (typo) is not, then we can raise a clear warning saying that `do_nomalize` isn't used anywhere. 
We can even push this a bit and do a quick levenshtein check conditioning a `perhaps you made a typo?` message. That might be too much.

For the decorator option to mark args as deprecated, LGTM too! Less loc, and forces us to write a version number. I'd say the "`from`" is not necessarily useful as we want to encourage using the latest version, imo. 
Thanks for looking into this @qubvel! Kind of related is https://github.com/huggingface/transformers/pull/30511 that I want to merge ASAP, didn't have time to get back to it with but it's closely connected to this problem.

Reply:
Thanks for looking into this @molbap 

Probably we can use a function signature in decorator to filter not-used kwargs instead of passing them.
What do you think about the following?

```python
import functools
import inspect
import warnings
import numpy as np


def filter_not_used_arguments(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):

        sig = inspect.signature(func)
        valid_params = sig.parameters.keys()
        valid_kwargs = {k: v for k, v in kwargs.items() if k in valid_params}
        invalid_params = {k: v for k, v in kwargs.items() if k not in valid_params}
        
        if invalid_params:
            warnings.warn(f"Unused named arguments: {', '.join(invalid_params.keys())}")
        
        return func(*args, **valid_kwargs)
    
    return wrapper


class ImageProcessor:

    @filter_not_used_arguments
    def preprocess(self, image, do_normalize=False, do_reduce_labels=False):
        pass


image_processor = ImageProcessor()
image = np.ones((100, 100, 3))

# passing invalid  `do_nomalize` instead of `do_normalize`
image_processor.preprocess(image, do_nomalize=True, do_reduce_labels=True)  
# UserWarning: Unused named arguments: do_nomalize
```

1. This will simplify validation, we will not need to store the list of available arguments
2. Make validation method-specific instead of class-specific
3. While this is backward compatible and we do not raise an error here, IDE will help users and force users to update parameters if no **kwargs is specified for the function
<img width="808" alt="Screenshot 2024-05-20 at 10 36 32" src="https://github.com/huggingface/transformers/assets/31920396/3edba496-39d5-403c-9fe6-cb27c556e71b">

Reply:
That looks like a great solution :) looks very good to me